frontend/src/session/websocket-provider.tsx

```diff
@@ -0,0 +1,71 @@
+import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
+import { Connection, Message, MessageType, WebsocketContext } from './websocket-context';
+
+export interface WebsocketProviderProps {
+  children: React.ReactNode;
+}
+
+const WS_URL = process.env.WS_URL || 'ws://localhost:8250/ws';
+const heartbeatInterval = 10000;
+const ping = JSON.stringify({ type: MessageType.PING });
+
+export const WebsocketProvider = ({ children }: WebsocketProviderProps) => {
```

Comment:
Please could you explain to me what the websocket context and provider are doing?

Reply:
Discussing offline